File: frontend/src/lib/tagRegistry.ts
/** The built-in tag registry — mirrors `bedwars_parser/tag_registry.py`.
 *
 * Colors and default binds are duplicated by hand (no cross-language codegen
 * here); if you change one side, change the other. This file only supplies
 * DEFAULTS for the Settings "quick rebind" rows — the actual source of truth
 * for a tag's current binding is the `keybind_map` the server returns, and
 * for its display label is `Store.list_tags()` (a renamed registry tag no
 * longer matches `registryEntryByLabel` here, same as the Python side).
 */

export interface TagRegistryEntry {
  id: string;
  label: string;
  defaultBind: string;
  color: string;
}

// CTRL+ALT combos, NOT bare F-keys: RegisterHotKey takes a binding
// EXCLUSIVELY, and bare F6-F9 stole a user's Medal clip hotkey.
// MUST match bedwars_parser/tag_registry.py — tests/test_tag_registry_sync.py
// fails the build if these two drift apart again.
export const TAG_REGISTRY: TagRegistryEntry[] = [
  { id: "my_mistake", label: "my mistake", defaultBind: "CTRL+ALT+F6", color: "#a371f7" },
  { id: "teammate_diff", label: "teammate diff", defaultBind: "CTRL+ALT+F7", color: "#58a6ff" },
  { id: "sweats", label: "sweats", defaultBind: "CTRL+ALT+F8", color: "#d29922" },
  { id: "cheater", label: "cheater", defaultBind: "CTRL+ALT+F9", color: "#ff7b72" },
];

export function registryEntryByLabel(label: string): TagRegistryEntry | undefined {
  return TAG_REGISTRY.find((e) => e.label === label);
}

// The four registry colors above, plus a handful more — shared by the
// Settings tag colour picker and any other swatch picker. Duplicates across
// tags are allowed; this list is not deduped against tags in use.
export const TAG_PALETTE: string[] = [
  "#a371f7",
  "#58a6ff",
  "#d29922",
  "#ff7b72",
  "#22c55e",
  "#ec4899",
  "#14b8a6",
  "#f97316",
  "#94a3b8",
];

File: bedwars_parser/tag_registry.py
"""The built-in tag registry — stable ids, labels, suggested default
keybinds, and colors for the four tags the app ships with.

This is NOT the source of truth for which tags exist — the ``tags`` SQLite
table is (the user can rename, delete, and create arbitrary tags there, and
everything keeps working for those; see ``keybind.py``, which binds on tag
LABEL, not on this registry's ``id``). This module is the source of truth for
the DEFAULT experience:

* ``db.py`` seeds a fresh database's tags AND its default keybind map from
  here, so keybind tagging works out of the box instead of starting empty.
* ``overlay.py`` looks a pressed tag's label up here to color its
  confirmation popup per-category, instead of a flat "green/amber" for every
  tag.
* The Settings "quick rebind" row per tag (frontend) shows this registry's
  color swatch and default binding as its starting point.

A tag NOT in the registry (the user renamed a registry tag, or created a new
one) still works everywhere — it just falls back to the generic action-based
styling instead of a per-tag color. Renaming a registry tag also detaches it
from the registry (lookup is by current label), which is correct: the
registry describes the DEFAULT four, not whatever the user calls them now.

Keep ``frontend/src/lib/tagRegistry.ts`` in sync by hand — colors and default
binds are duplicated there (no cross-language codegen here) — comments in
both files point at each other.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class TagRegistryEntry:
    id: str            # stable slug — never shown, never renamed
    label: str          # default display name == the seeded Store tag name
    default_bind: str   # a keybind.parse_binding()-valid key string
    color: str           # hex; matches Store._DEFAULT_TAGS


# CTRL+ALT combos, not bare F-keys.
#
# RegisterHotKey takes EXCLUSIVE global ownership of a binding: while the
# tracker runs, the key stops reaching every other application. The original
# bare F6-F9 defaults broke a user's Medal clip hotkey for exactly that
# reason. Bare F-keys are still bindable by hand — they are just not what we
# take from people by default.
TAG_REGISTRY: tuple[TagRegistryEntry, ...] = (
    TagRegistryEntry("my_mistake", "my mistake", "CTRL+ALT+F6", "#a371f7"),
    TagRegistryEntry("teammate_diff", "teammate diff", "CTRL+ALT+F7", "#58a6ff"),
    TagRegistryEntry("sweats", "sweats", "CTRL+ALT+F8", "#d29922"),
    TagRegistryEntry("cheater", "cheater", "CTRL+ALT+F9", "#ff7b72"),
)


def by_label(label: str) -> "TagRegistryEntry | None":
    """The registry entry whose label currently matches, or None — a renamed
    or user-created tag correctly finds nothing here."""
    for entry in TAG_REGISTRY:
        if entry.label == label:
            return entry
    return None


def default_keymap() -> dict[str, str]:
    """The keybind_map a fresh install starts with: each entry's
    default_bind -> label. Every key is valid input to
    ``keybind.parse_binding`` (asserted in tests, since a bad default here
    would silently fail to register on every fresh install)."""
    return {entry.default_bind: entry.label for entry in TAG_REGISTRY}
